File: src/rust/cpu2/global_pointers.rs
#![allow(non_upper_case_globals)]

use cpu2::cpu::reg128;

pub const reg8: *mut u8 = 64 as *mut u8;
pub const reg16: *mut u16 = 64 as *mut u16;
pub const reg8s: *mut i8 = 64 as *mut i8;
pub const reg16s: *mut i16 = 64 as *mut i16;
pub const reg32: *mut i32 = 64 as *mut i32;

pub const last_op1: *mut i32 = 96 as *mut i32;

pub const last_op_size: *mut i32 = 104 as *mut i32;

pub const last_result: *mut i32 = 112 as *mut i32;
pub const flags_changed: *mut i32 = 116 as *mut i32;
pub const flags: *mut i32 = 120 as *mut i32;

pub const page_fault: *mut bool = 540 as *mut bool;
pub const page_fault_error_code: *mut i32 = 544 as *mut i32;

pub const instruction_pointer: *mut i32 = 556 as *mut i32;
pub const previous_ip: *mut i32 = 560 as *mut i32;
pub const idtr_size: *mut i32 = 564 as *mut i32;
pub const idtr_offset: *mut i32 = 568 as *mut i32;
pub const gdtr_size: *mut i32 = 572 as *mut i32;
pub const gdtr_offset: *mut i32 = 576 as *mut i32;
pub const cr: *mut i32 = 580 as *mut i32;
pub const cpl: *mut u8 = 612 as *mut u8;
pub const in_hlt: *mut bool = 616 as *mut bool;
pub const last_virt_eip: *mut i32 = 620 as *mut i32;
pub const eip_phys: *mut i32 = 624 as *mut i32;
pub const last_virt_esp: *mut i32 = 628 as *mut i32;
pub const esp_phys: *mut i32 = 632 as *mut i32;
pub const sysenter_cs: *mut i32 = 636 as *mut i32;
pub const sysenter_esp: *mut i32 = 640 as *mut i32;
pub const sysenter_eip: *mut i32 = 644 as *mut i32;
pub const prefixes: *mut u8 = 648 as *mut u8;
pub const timestamp_counter: *mut u32 = 664 as *mut u32;
pub const sreg: *mut u16 = 668 as *mut u16;
pub const dreg: *mut i32 = 684 as *mut i32;

pub const segment_is_null: *mut bool = 724 as *mut bool;
pub const segment_offsets: *mut i32 = 736 as *mut i32;
pub const segment_limits: *mut u32 = 768 as *mut u32;
pub const protected_mode: *mut bool = 800 as *mut bool;
pub const is_32: *mut bool = 804 as *mut bool;
pub const stack_size_32: *mut bool = 808 as *mut bool;
pub const memory_size: *mut u32 = 812 as *mut u32;
pub const fpu_stack_empty: *mut i32 = 816 as *mut i32;
pub const mxcsr: *mut i32 = 824 as *mut i32;
// gap
pub const reg_xmm: *mut reg128 = 832 as *mut reg128;
pub const current_tsc: *mut u64 = 960 as *mut u64;
pub const fpu_st: *mut f64 = 968 as *mut f64;
pub const fpu_st8: *mut u8 = 968 as *mut u8;
pub const fpu_st32: *mut i32 = 968 as *mut i32;
pub const fpu_stack_ptr: *mut u32 = 1032 as *mut u32;
pub const fpu_control_word: *mut i32 = 1036 as *mut i32;
pub const fpu_status_word: *mut i32 = 1040 as *mut i32;
pub const fpu_opcode: *mut i32 = 1044 as *mut i32;
pub const fpu_ip: *mut i32 = 1048 as *mut i32;
pub const fpu_ip_selector: *mut i32 = 1052 as *mut i32;
pub const fpu_dp: *mut i32 = 1056 as *mut i32;
pub const fpu_dp_selector: *mut i32 = 1060 as *mut i32;
pub const reg_mmx: *mut u64 = 1064 as *mut u64;
pub const tss_size_32: *mut bool = 1128 as *mut bool;
pub const fxsave_store_fpu_mask: *mut u8 = 1132 as *mut u8;
pub const opstats_buffer: *mut u32 = 0x08000 as *mut u32;
pub const opstats_compiled_buffer: *mut u32 = 0x10000 as *mut u32;
pub const opstats_jit_exit_buffer: *mut u32 = 0x18000 as *mut u32;
pub const opstats_unguarded_register_buffer: *mut u32 = 0x20000 as *mut u32;
pub const opstats_wasm_size: *mut u32 = 0x28000 as *mut u32;
pub const tlb_data: *mut i32 = 0x400000 as *mut i32;

File: src/native/global_pointers.c
#include <math.h>
#include <stdbool.h>
#include <stdint.h>

#include "const.h"
#include "shared.h"

uint8_t* const reg8 = (uint8_t* const) 4;
uint16_t* const reg16 = (uint16_t* const) 4;
int8_t* const reg8s = (int8_t* const) 4;
int16_t* const reg16s = (int16_t* const) 4;
int32_t* const reg32s = (int32_t* const) 4;

int32_t* const last_op1 = (int32_t* const) 512;
int32_t* const last_op2 = (int32_t* const) 516;
int32_t* const last_op_size = (int32_t* const) 520;
int32_t* const last_add_result = (int32_t* const) 524;
int32_t* const last_result = (int32_t* const) 528;
int32_t* const flags_changed = (int32_t* const) 532;
int32_t* const flags = (int32_t* const) 536;

bool* const page_fault = (bool* const) 540;

// gap

bool* const a20_enabled = (bool* const) 552;
int32_t* const instruction_pointer = (int32_t* const) 556;
int32_t* const previous_ip = (int32_t* const) 560;

int32_t* const idtr_size = (int32_t* const) 564;
int32_t* const idtr_offset = (int32_t* const) 568;
int32_t* const gdtr_size = (int32_t* const) 572;
int32_t* const gdtr_offset = (int32_t* const) 576;
int32_t* const cr = (int32_t* const) 580; // length 32

uint8_t* const cpl = (uint8_t* const) 612;
bool* const in_hlt = (bool* const) 616;
int32_t* const last_virt_eip = (int32_t* const) 620;
int32_t* const eip_phys = (int32_t* const) 624;
int32_t* const last_virt_esp = (int32_t* const) 628;
int32_t* const esp_phys = (int32_t* const) 632;
int32_t* const sysenter_cs = (int32_t* const) 636;
int32_t* const sysenter_esp = (int32_t* const) 640;
int32_t* const sysenter_eip = (int32_t* const) 644;
uint8_t* const prefixes = (uint8_t* const) 648;
// gap
uint32_t* const timestamp_counter = (uint32_t* const) 664;

uint16_t* const sreg = (uint16_t* const) 668;
int32_t* const dreg = (int32_t* const) 684; // length 32
int32_t* const fw_value = (int32_t* const) 720;
bool* const segment_is_null = (bool* const) 724; // length 8
int32_t* const segment_offsets = (int32_t* const) 736; // length 32
uint32_t* const segment_limits = (uint32_t* const) 768; // length 32

bool* const protected_mode = (bool* const) 800;
bool* const is_32 = (bool* const) 804;
bool* const stack_size_32 = (bool* const) 808;
uint32_t* const memory_size = (uint32_t* const) 812;
int32_t* const fpu_stack_empty = (int32_t* const) 816;

// gap

int32_t* const mxcsr = (int32_t* const) 824;

union reg128* const reg_xmm = (union reg128* const) 828; // length 128

uint64_t* const current_tsc = (uint64_t* const) 956;

double_t* const fpu_st = (double_t* const) 968; // length 64
uint8_t* const fpu_st8 = (uint8_t* const) 968;
int32_t* const fpu_st32 = (int32_t* const) 968;

uint32_t* const fpu_stack_ptr = (uint32_t* const) 1032;
int32_t* const fpu_control_word = (int32_t* const) 1036;
int32_t* const fpu_status_word = (int32_t* const) 1040;
int32_t* const fpu_opcode = (int32_t* const) 1044;
int32_t* const fpu_ip = (int32_t* const) 1048;
int32_t* const fpu_ip_selector = (int32_t* const) 1052;
int32_t* const fpu_dp = (int32_t* const) 1056;
int32_t* const fpu_dp_selector = (int32_t* const) 1060;

union reg64* const reg_mmx = (union reg64* const) 1064; // length 64

extern bool* const tss_size_32 = (bool* const) 1128;

// gap

uint32_t* const opstats_buffer = (uint32_t* const) 0x1000; // length 0x400
uint32_t* const opstats_buffer_0f = (uint32_t* const) 0x1400; // length 0x400

// gap

int32_t* const tlb_data = (int32_t* const) 0x400000; // length 0x100000*4
